### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by Tiberriver256** [OData/AspNetCoreOData#363]
I love OData and the AspNetCore implementation is so powerful. I'd like to recommend the use of this project to enterprise clients but the lack of documentation makes this an almost impossible learning curve (despite the fact that OData is actually quite simple).

I recently went through [this course](https://app.pluralsight.com/library/courses/building-consistent-restful-api-odata-v4-asp-dot-net-core) on Pluralsight and wanted to put together a check-list of things that should be documented somewhere.

I would love to help out with PRs on some of these if you guys are open to having someone help out. Even just as PRs to the docs folder would be helpful for people searching for documentation. If the documentation for some of these things exists somewhere already (i.e., in a blog post), I think some amount of effort should be made to:

1. Move that documentation out of a blog post to here in order to keep it centralized
2. Ensure it's updated to reflect v8

## Documentation Needs

- [ ] Attribute-based Routing
- [ ] Convention-based Routing
- [ ] Defining an Entity Data Model (EDM)
- [ ] Defining Create/Read/Update/Delete methods for an entity set
- [ ] Working with properties and raw property values
- [ ] Working with navigation properties
- [ ] Querying data (should cover `EnableQueryAttribute`, `ODataQueryOptions`, and global query options)
- [ ] Working with containment navigation
- [ ] Working with Functions
- [ ] Working with Actions
- [ ] Working with Singletons
- [ ] Working with Derived Types
- [ ] Working with Open Types
- [ ] Working with Batch Processing
- [ ] Configuring OpenApi/Swagger
- [ ] Other OData Learning - I think we should assume people that come here are familiar with how OData works but if they are not we should direct them to some basic documentation on how OData works. I feel like this is already done well enough elsewhere we shouldn't need to re-document it here.
- [ ] Setting up [authentication/authorization information in your Edm](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Authorization.V1.md)

## Related Requests

#332 
#23
#63 
#66

**Comment by Tiberriver256** [comment]
@asanyaga - Hey just wanted to follow up and ask, are you guys open to community submitted PRs for this? I'd like to try and tackle a document on routing.

**Comment by asanyaga** [comment]
Hey @Tiberriver256 - yes we are open to community contributions

**Comment by asanyaga** [comment]
@Tiberriver256 would also like to hear your thoughts on what are the opportunities for improvement in the OData docs

**Comment by Tiberriver256** [comment]
Awesome @asanyaga! I'll try and get a start on this over the weekend.

I'd be open to reviewing anything if you want me to point me in a direction. Although I haven't used the .NET Framework version so I wouldn't be much help [with these docs](https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/). I would love to see something similar set up for AspNetCoreOdata on docs.microsoft.com though (or if there's already something there you could point me to that would be wonderful).

The only documentation I'm aware of for AspNetCoreOData are the below resources so I'll give you my quick feedback on those:

* The [docs folder](https://github.com/OData/AspNetCoreOData/tree/master/docs) in this repo - This is very sparse and outdated
* The [readme](https://github.com/OData/AspNetCoreOData/blob/master/README.md) - This has a nice getting started guide and some links to example projects in leu of documentation
* The [odata blog](https://devblogs.microsoft.com/odata/) - This is a great feed to follow for new updates but not useful as documentation since

**Comment by Tiberriver256** [comment]
Also just noticed this one article that seems applies to AspNetCore OData but is put in with all the .NET Framework docs:
https://docs.microsoft.com/en-us/odata/webapi/nested-paths-routing

The link to the `Automatic nested paths` guide is also a dead link.

**Comment by Tiberriver256** [comment]
@xuzhg - I just realized the Microsoft docs are actually driven from [this repo](https://github.com/MicrosoftDocs/OData-docs). Should I be doing these docs over there?

**Comment by xuzhg** [comment]
@Tiberriver256 First, Thanks for your contribution. I'd love to see more from you and the community. :)

At the current stage, maybe it's easy for the community to learn the docs at the same repository. So, let's focus on the docs folder in this repo.

**Comment by asanyaga** [comment]
> @Tiberriver256 First, Thanks for your contribution. I'd love to see more from you and the community. :)
> 
> At the current stage, maybe it's easy for the community to learn the docs at the same repository. So, let's focus on the docs folder in this repo.

@Tiberriver256 thanks for the contribution. I'm also reaching out to the content team to find out what the guidelines are for community contributions to the ms docs repo. cc @xuzhg

**Comment by LarsKemmann** [comment]
![image](https://user-images.githubusercontent.com/5834429/199984167-5059b253-0dbd-48c3-a017-e5bb4acc1800.png)

Seriously, OData used to be _very_ well documented by Microsoft. How is it that this crucial component of the ASP.NET stack now has so little meaningful documentation? If my scenario doesn't _exactly_ fit within the "happy path" of exposing an Entity Framework model, there's zero support outside of some (undocumented) samples. I'm left with trial-and-error coding, a paltry bit of guidance via IntelliSense, and submitting comments on issues that probably won't be taken seriously.

@xuzhg:
> At the current stage, maybe it's easy for the community to learn the docs at the same repository. So, let's focus on the docs folder in this repo.

The past year has shown that this was the wrong way to go. ASP.NET Core team members may not experience this themselves, but the community _heavily_ relies on the [formal  ASP.NET Core docs](https://learn.microsoft.com/en-us/aspnet/core/?view=aspnetcore-6.0). The reality is that by keeping the documentation in this repository (where it has received almost no attention over the past year since this issue was opened), **the OData team has taken the easy way out**—with fewer eyes on the documentation, there are fewer people who will take the time to find their way here to complain about how absurdly bad it is.

I should be able to expect a level of proper documentation comparable to what [the SignalR team](https://learn.microsoft.com/en-us/aspnet/core/signalr/introduction?view=aspnetcore-6.0) provides.

**Comment by julealgon** [comment]
> @xuzhg:
> 
> > At the current stage, maybe it's easy for the community to learn the docs at the same repository. So, let's focus on the docs folder in this repo.
> 
> The past year has shown that this was the wrong way to go. ASP.NET Core team members may not experience this themselves, but the community _heavily_ relies on the [formal ASP.NET Core docs](https://learn.microsoft.com/en-us/aspnet/core/?view=aspnetcore-6.0). 

I agree with this. While it is nice to have documentation closer to the code (in the repo), the fact that MSDN docs only have information regarding non-Core OData is jarring right now, especially because some of that documentation still applies to AspNetCoreOData but this is not mentioned there.

If the team really wants to maintain the docs in the repo, there should be a very clear redirection from MSDN somewhere so that people landing there first can easily find the docs here.

I'll apologize upfront for the perhaps blunt comment, but to me AspNetCoreOData's documentation right now feels amateurish to put it gently: I know a lot of the new stuff is mentioned in blog posts linked in the readme, but blog posts should never be the source of truth for a library IMHO. Instead, they should be used to hype up some new features, show some examples, as an introduction, but still link to a "formal" documentation somewhere that has full detail about the functionality.

Couple this documentation problem with the overall lack of awareness and publicity OData gets, and suddenly you face an uphill battle trying to convince other developers to try OData out in real world scenarios. The library not only needs to be well supported, but it also needs to _appear_ well supported.

All in all, the poor documentation that exists today is scattered all over the place, in blog posts, sometimes in issues/discussions, in samples, most of it is still "hidden" as part of the original OData docs in MSDN.

**Comment by dnperfors** [comment]
> I know a lot of the new stuff is mentioned in blog posts linked in the readme, but blog posts should never be the source of truth for a library IMHO.

Excellent point, besides that, since those blog posts a lot has been changed and some of them aren't even working anymore.

My way of finding out how things need to be done right now is mostly looking in the source code and stepping through the implementation, which is not a pleasant thing to do when you want to deliver value for your customers.
